File: scripts/ci-upload-windows.py
#!/usr/bin/env python3
"""Upload Windows build artifacts from GitHub Actions runner to Aliyun OSS.

Env vars required:
  OSS_ACCESS_KEY_ID
  OSS_ACCESS_KEY_SECRET

Usage:
  python scripts/ci-upload-windows.py <version>
"""

import os
import sys
from pathlib import Path

import oss2

BUCKET_NAME = "lotus-releases"
ENDPOINT = "https://oss-cn-beijing.aliyuncs.com"
OSS_PREFIX = "aijia"


def upload(bucket, local, key):
    size = os.path.getsize(local)
    print(f"[upload] {os.path.basename(local)} ({size / 1024 / 1024:.1f}MB) -> {key}")
    oss2.resumable_upload(
        bucket, key, str(local),
        multipart_threshold=10 * 1024 * 1024,
        part_size=5 * 1024 * 1024,
        num_threads=4,
    )


def main():
    if len(sys.argv) < 2:
        print("Usage: python ci-upload-windows.py <version>")
        sys.exit(1)
    version = sys.argv[1].lstrip("v")

    key_id = os.environ.get("OSS_ACCESS_KEY_ID", "")
    key_secret = os.environ.get("OSS_ACCESS_KEY_SECRET", "")
    if not key_id or not key_secret:
        print("[error] OSS_ACCESS_KEY_ID / OSS_ACCESS_KEY_SECRET not set")
        sys.exit(1)

    project_root = Path(__file__).resolve().parent.parent
    nsis_dir = project_root / "src-tauri" / "target" / "release" / "bundle" / "nsis"
    exe = nsis_dir / f"AIjia_{version}_x64-setup.exe"
    sig = nsis_dir / f"AIjia_{version}_x64-setup.exe.sig"

    if not exe.exists():
        print(f"[error] Installer not found: {exe}")
        print(f"  nsis dir contents: {[p.name for p in nsis_dir.iterdir()] if nsis_dir.exists() else '(missing)'}")
        sys.exit(1)

    auth = oss2.Auth(key_id, key_secret)
    bucket = oss2.Bucket(auth, ENDPOINT, BUCKET_NAME)

    exe_key = f"{OSS_PREFIX}/v{version}/AIjia_{version}_x64-setup.exe"
    upload(bucket, exe, exe_key)

    latest_key = f"{OSS_PREFIX}/latest/windows-x64"
    bucket.copy_object(
        BUCKET_NAME, exe_key, latest_key,
        headers={
            "x-oss-metadata-directive": "REPLACE",
            "Content-Type": "application/octet-stream",
            "Content-Disposition": f'attachment; filename="AIjia_{version}_x64-setup.exe"',
        },
    )
    print(f"  -> latest: {latest_key}")

    if sig.exists():
        sig_key = exe_key + ".sig"
        print(f"[upload] {sig.name} -> {sig_key}")
        bucket.put_object_from_file(sig_key, str(sig))
    else:
        print(f"[warn] Signature not found: {sig} -- updater will not work without it")
        sys.exit(1)

    print(f"\n[ok] Windows v{version} uploaded to OSS")


if __name__ == "__main__":
    main()

File: scripts/ci-upload-windows.mjs
#!/usr/bin/env node
// Upload signed Windows installer + Tauri updater .sig to Aliyun OSS.
//
// Pure Node — no Python dependency. Uses the official `ali-oss` SDK.
//
// Usage:
//   node scripts/ci-upload-windows.mjs <version> <release|beta> <exe-path>
//
// Required env vars:
//   OSS_ACCESS_KEY_ID
//   OSS_ACCESS_KEY_SECRET
//
// The .sig file is expected at <exe-path>.sig and uploaded alongside.
// For release type, also copies to aijia/latest/windows-x64.

import { readFileSync, statSync, existsSync } from 'node:fs'
import { basename } from 'node:path'
import { spawnSync } from 'node:child_process'

const BUCKET = 'lotus-releases'
const REGION = 'oss-cn-beijing'
const PREFIX = 'aijia'

const [, , version, releaseType, exePath] = process.argv
if (!version || !releaseType || !exePath) {
  console.error('Usage: node scripts/ci-upload-windows.mjs <version> <beta|release> <exe-path>')
  process.exit(1)
}
if (releaseType !== 'beta' && releaseType !== 'release') {
  console.error(`ERROR: release type must be 'beta' or 'release', got: ${releaseType}`)
  process.exit(1)
}
if (!existsSync(exePath)) {
  console.error(`ERROR: exe not found: ${exePath}`)
  process.exit(1)
}
const sigPath = `${exePath}.sig`
if (!existsSync(sigPath)) {
  console.error(`ERROR: .sig not found: ${sigPath}`)
  console.error('  Generate it with: tauri signer sign -k ~/.tauri/aijia.key <exe>')
  process.exit(1)
}

const keyId = process.env.OSS_ACCESS_KEY_ID
const keySecret = process.env.OSS_ACCESS_KEY_SECRET
if (!keyId || !keySecret) {
  console.error('ERROR: OSS_ACCESS_KEY_ID / OSS_ACCESS_KEY_SECRET not set')
  process.exit(1)
}

// Lazy-install ali-oss into a local node_modules if not present.
// Avoids polluting the repo's package.json with a release-only dep.
let OSS
try {
  ({ default: OSS } = await import('ali-oss'))
} catch {
  console.log('[setup] ali-oss not installed in this project — installing locally...')
  const r = spawnSync('npm', ['install', '--no-save', '--no-audit', '--no-fund', 'ali-oss@6'], {
    stdio: 'inherit',
    shell: process.platform === 'win32',
  })
  if (r.status !== 0) {
    console.error('ERROR: failed to install ali-oss')
    process.exit(1)
  }
  ;({ default: OSS } = await import('ali-oss'))
}

const client = new OSS({
  region: REGION,
  bucket: BUCKET,
  accessKeyId: keyId,
  accessKeySecret: keySecret,
  // 1 hour timeout for large uploads on slow networks
  timeout: 3600 * 1000,
  secure: true,
})

const ossPrefix = releaseType === 'beta'
  ? `${PREFIX}/beta/v${version}`
  : `${PREFIX}/v${version}`
const exeKey = `${ossPrefix}/AIjia_${version}_x64-setup.exe`
const sigKey = `${exeKey}.sig`

async function uploadOne(localPath, remoteKey) {
  const size = statSync(localPath).size
  const sizeMB = (size / 1024 / 1024).toFixed(1)
  console.log(`[upload] ${basename(localPath)} (${sizeMB} MB) -> ${remoteKey}`)

  if (size > 10 * 1024 * 1024) {
    // Multipart for large files (resumable, parallel)
    await client.multipartUpload(remoteKey, localPath, {
      partSize: 5 * 1024 * 1024,
      parallel: 4,
      progress: (p) => process.stdout.write(`\r  progress ${(p * 100).toFixed(0)}%   `),
    })
    process.stdout.write('\n')
  } else {
    await client.put(remoteKey, localPath)
  }
}

try {
  await uploadOne(exePath, exeKey)
  await uploadOne(sigPath, sigKey)

  if (releaseType === 'release') {
    const latestKey = `${PREFIX}/latest/windows-x64`
    console.log(`[copy ] ${exeKey} -> ${latestKey}`)
    await client.copy(latestKey, exeKey, {
      headers: {
        'x-oss-metadata-directive': 'REPLACE',
        'Content-Type': 'application/octet-stream',
        'Content-Disposition': `attachment; filename="AIjia_${version}_x64-setup.exe"`,
      },
    })
  }

  console.log(`\n[ok] Windows v${version} (${releaseType}) uploaded to OSS`)
  console.log(`     https://lotus.renlijia.com/${exeKey}`)
} catch (err) {
  console.error(`\nERROR: upload failed: ${err.message}`)
  if (err.code) console.error(`  OSS error code: ${err.code}`)
  if (err.requestId) console.error(`  request id: ${err.requestId}`)
  process.exit(1)
}
